### ISSUE TITLE
block is nil when it shouldn't be

### ISSUE THREAD

**Issue by cj** [opal/opal#895]
To replicate checkout https://github.com/wedgeio/wedge `bundle install` and run `bundle exec rake` you'll see it print out a statement client and server side.  You'll see `"name"=>"ready", "block"=>nil` on the client and  `:name=>:ready, :block=>#<Proc:` on the server side.  They should both be procs.

If you remove https://github.com/wedgeio/wedge/blob/master/lib/wedge/opal.rb#L31 it will work and they both have procs.  If you change that line to just `@result << Opal.original_compile("require '#{self.file}'")`, it will fail again.  https://github.com/wedgeio/wedge/blob/master/playground/app/components/uploader.rb#L33-L40 is where the block is defined.

If you also just remove https://github.com/wedgeio/wedge/blob/master/Rakefile#L8 it will work and both will have a proc.  It seems like requiring the same file a bunch of times is some how making blocks nil. 

The on method it is calling is https://github.com/wedgeio/wedge/blob/master/lib/wedge/component.rb#L120 if you put a `puts block` in there you'll see that it's nil too so it's not the value being made nil in the app.

**Comment by meh** [comment]
We're aware of this bug, I hit it various times, we have absolutely no clue why it's happening :rage4:

**Comment by elia** [comment]
I hope jsilnt will help with this

**Comment by meh** [comment]
@elia that's also my hope, it must be some mistakenly global var fucking everything up.

**Comment by cj** [comment]
:sob:

**Comment by vais** [comment]
@cj can't get it to work at all (tried on Windows and on Linux). Here's what I'm getting in both cases:
```
$ bundle exec rake
rake aborted!
KeyError: key not found
/home/vais/Desktop/wedge/playground/app/config/variables.rb:14:in `fetch'
/home/vais/Desktop/wedge/playground/app/config/variables.rb:14:in `<top (required)>'
/home/vais/Desktop/wedge/playground/app/config/boot.rb:3:in `require'
/home/vais/Desktop/wedge/playground/app/config/boot.rb:3:in `<top (required)>'
/home/vais/Desktop/wedge/Rakefile:8:in `require'
/home/vais/Desktop/wedge/Rakefile:8:in `<top (required)>'
(See full trace by running task with --trace)
```
This makes it that much more difficult to work on the issue, cause I need to understand everything about wedge first, before I can even start looking at the Opal problem. Do you think you can reduce this issue to the simplest reproducible case that has nothing to do with wedge?

**Comment by cj** [comment]
@vais it was trying to fetch some required variables, from the variables.rb file.  I've made them not required now so it won't throw that error.  If you run `bundle exec rake` again it will work.

As for the code it should be pretty straight forward, not that much 'wedge' specific things.  The block that is nil client side is this one:

https://github.com/wedgeio/wedge/blob/master/playground/app/components/uploader.rb#L33-L40

Which then just calls this method:

https://github.com/wedgeio/wedge/blob/master/lib/wedge/component.rb#L120

If you `puts block` just above the case statement you'll see i't nil there too.

There's also some additional code compiled here:

https://github.com/wedgeio/wedge/blob/master/lib/wedge/opal.rb#L19

**Comment by vais** [comment]
@cj, no problem, tried again, now it's a different error:
```
$ bundle exec rake
rake aborted!
NoMethodError: undefined method `include' for Wedge::Component:Class
/home/vais/Desktop/wedge/lib/wedge/component.rb:137:in `method_missing'
/home/vais/Desktop/wedge/lib/wedge/config.rb:53:in `plugin'
/home/vais/Desktop/wedge/lib/roda/plugins/wedge.rb:16:in `block (2 levels) in configure'
/home/vais/Desktop/wedge/lib/roda/plugins/wedge.rb:16:in `each'
/home/vais/Desktop/wedge/lib/roda/plugins/wedge.rb:16:in `block in configure'
/home/vais/Desktop/wedge/lib/roda/plugins/wedge.rb:13:in `each'
/home/vais/Desktop/wedge/lib/roda/plugins/wedge.rb:13:in `configure'
/home/vais/Desktop/wedge/playground/app/app.rb:16:in `<class:Playground>'
/home/vais/Desktop/wedge/playground/app/app.rb:1:in `<top (required)>'
/home/vais/Desktop/wedge/lib/wedge/require.rb:4:in `require'
/home/vais/Desktop/wedge/lib/wedge/require.rb:4:in `require'
/home/vais/Desktop/wedge/playground/app/config/boot.rb:16:in `<top (required)>'
/home/vais/Desktop/wedge/Rakefile:8:in `require'
/home/vais/Desktop/wedge/Rakefile:8:in `<top (required)>'
(See full trace by running task with --trace)
```
Again, the first thing I would have to do once `bundle exec rake` works is to create the absolute minimum test case that reproduces the problem. Without that, there's a snowball's chance in hell of fixing the problem. If you could come up with such a minimum required example based on your experience with this issue, it would help move things along towards a potential resolution.

**Comment by cj** [comment]
@vais what version of ruby are you running?

Once this works it should be a minimal example.  The code getting executed is less than 30 lines.

**Comment by cj** [comment]
I updated the code, it shouldn't give you that error anymore.  What version of ruby was it though and was it windows or linux?

**Comment by vais** [comment]
@cj 
```
$ ruby -v
ruby 2.0.0p643 (2015-02-25 revision 49749) [x86_64-linux]
```

**Comment by cj** [comment]
@vais thank you.  I just installed that version and `bundle exec rake` worked just fine.  Is that the case for you now?

**Comment by vais** [comment]
@cj yep, it's working as of your last update. I'll see if I can wrap my head around the issue...

**Comment by cj** [comment]
@vais great!  thank you!  Almost 2am here but I'll be back on around 8-9am my time so in 6-7 hours.

**Comment by vais** [comment]
@cj the root cause of this behavior remains a mystery for now, but here's a work-around that fixes this problem for you in the meantime:

`playground/app/components/uploader.rb` 
```diff
-    on :ready do
+    on :ready, proc {
       button = dom.find("button")
       button.hide
       wedge_plugin(:uploader).button button, {
         wedge_name: :uploader,
         wedge_method: :add_file
       }
-    end
+    }
```
`lib/wedge/plugins/uploader.rb`
```diff
-      on :compile do
+      on :compile, proc {
         settings = Wedge.config.settings[:uploader]
         store[:settings] = settings.select do |k, v|
           client? ? %w(aws_access_key_id bucket).include?(k) : true
         end if settings
-      end
+      }
```
`lib/wedge/component.rb`
```diff
-      def wedge_on(*args, &block)
+      def wedge_on(*args)
+        block = args.pop
         case args.first.to_s
         when 'server'
           wedge_on_server(&block)
         when 'compile'
           wedge_config.on_compile << block unless RUBY_ENGINE == 'opal'
```
@meh, @elia, given the above work-around, it would appear that the issue is perhaps somewhere in the interaction between variable number of arguments and block passing?

**Comment by cj** [comment]
@vais thank you for taking a look.  I was aware of the work around, I just can't do it as we have it in a few large projects and that would mean changing it in 100 - 200 places.  For some reason this isn't an issue for them.  It just means at the moment I can't write a full test suit for wedge.

Hopefully if we all work together we can figure out the root cause.  If you need me to do anything just let me know.

**Comment by vais** [comment]
@cj the issue is resolved on the latest master.

**Comment by vais** [comment]
tldr: `..."block"=>#<Proc:0x5d4>...` in your test case under Opal :sparkles: :joy: :beers: :beer: :tada: :sparkles: 

```
$ bundle exec rake
--------------------------
Run specs in Opal
--------------------------
Thin web server (v1.6.3 codename Protein Powder)
Maximum connections set to 1024
Listening on localhost:9999, CTRL+C to stop
WARN: tilt autoloading 'tilt/erubis' in a non thread-safe way; explicit require 'tilt/erubis' suggested.
WARNING: Cannot handle dynamic require -- rspec/core/configuration:
WARNING: Cannot handle dynamic require -- rspec/core/configuration:533
WARNING: Cannot handle dynamic require -- rspec/core/configuration:857
WARNING: Cannot handle dynamic require -- rspec/core/configuration:1095
WARNING: Cannot handle dynamic require -- rspec/core/configuration:1095
WARNING: Cannot handle dynamic require -- rspec/core:
WARNING: Cannot handle dynamic require -- rspec/core:200
WARNING: Cannot handle dynamic require -- rspec/core:200
WARNING: LoadError: cannot load such file -- opal
{"uploader"=>{"browser_events"=>[{"name"=>"ready", "block"=>#<Proc:0x5d4>, "wedge_name"=>"uploader", "options"=>{}}], "object_events"=>{}, "on_count"=>1}}


Finished


4 examples, 0 failures (time taken: 0.026)
--------------------------
Run specs in normal ruby
--------------------------
/Users/vais/.rbenv/versions/2.2.0/bin/ruby -I/Users/vais/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib:/Users/vais/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib /Users/vais/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb
{"uploader"=>{"browser_events"=>[{:name=>:ready, :block=>#<Proc:0x007faf9433f680@/Users/vais/Desktop/wedge/playground/app/components/uploader.rb:33>, :wedge_name=>:uploader, :options=>{}}], "object_events"=>{}, "on_count"=>1}}
Run options: exclude {:slow=>true}
.......

Finished in 0.02612 seconds (files took 0.50483 seconds to load)
7 examples, 0 failures
```

**Comment by cj** [comment]
@vais awesome!!!  thank you! :smile:

**Comment by vais** [comment]
@cj the pleasure's all mine, so, you're welcome, and thanks for persevering!
